### ISSUE TITLE
Seq bytes __eq__ implementation forces non-string comparator to be bytes object

### ISSUE THREAD

**Issue by hypostulate** [biopython/biopython#3597]
### Setup

Biopython 1.79

### Expected behaviour

Previously, when working with a custom class as follows, all assertions in the test would pass:

```python
class SeqLike:
    def __init__(self, seq: str):
        self._seq = seq
    
    def __str__(self) -> str:
        return self._seq
    
    def __eq__(self, other: object) -> bool:
        # Primarily comparison would be `Union[str, Seq]` for this
        # example
        return self._seq == str(other)
    
def test_equality():
    from Bio.Seq import Seq
    seq = Seq("ATG")
    seq_like = SeqLike("ATG")
    assert seq_like == seq       # Uses SeqLike.__eq__ True
    assert seq == str(seq_like)  # Uses Seq.__eq__ True
    assert seq == seq_like       # Uses Seq.__eq__ False
```

### Actual behaviour

The third assertion (using `Seq.__eq__`) fails because we are actually making this comparison:  `b"ATG" == SeqLike("ATG")` and there the `SeqLike` is not cast to string at any point. In version 1.78, the `Seq.__eq__` function was simply `str(self) == str(other)` and the call to `str(...)` would route through `SeqLike.__str__` and behave appropriately, that gave us the hook to control the right-sided equality operation. Without the `str(other)` call in the current implementation, we have no handle to change the behavior of the right-handed equality operation (when our object is on the right of `==`). 

### Steps to reproduce

Biopython 1.79, run the example above.

### Suggested improvement

In `Seq.__eq__` replace:

```python
        if isinstance(other, _SeqAbstractBaseClass):
            return self._data == other._data
        elif isinstance(other, str):
            return self._data == other.encode("ASCII")
        else:
            return self._data == other
```

with

```python
        if isinstance(other, _SeqAbstractBaseClass):
            return self._data == other._data
        elif isinstance(other, bytes):
            return self._data == other
        elif isinstance(other, bytearray):
            return self._data == other
        else:
            return self._data == str(other).encode("ASCII")
```

This should keep the behavior the same for `byte` and `bytearray` direct equality, but it implements a catchall attempt at `str` casting the other object and gives back control.

**Comment by hypostulate** [comment]
This error can actually be resolved from the `SeqLike` class without modification to `Seq.__eq__`:

```python
class SeqLike:
    def __init__(self, seq: str):
        self._seq = seq
    
    def __str__(self) -> str:
        return self._seq
    
    def __eq__(self, other: object) -> bool:
        # Primarily comparison would be `Union[str, Seq]` for this
        # example        
        # Adding specific logic for `bytes` and `bytearray` allows handling both equality configurations
        if isinstance(other, bytes):
            return bool(self._seq == other.decode())
        elif isinstance(other, bytearray):
            return bool(self._seq == other.decode())
        return self._seq == str(other)
    
def test_equality():
    from Bio.Seq import Seq
    seq = Seq("ATG")
    seq_like = SeqLike("ATG")
    assert seq_like == seq
    assert seq == str(seq_like)
    assert seq == seq_like
```

**Comment by peterjc** [comment]
Anything to add @mdehoon ?

**Comment by mdehoon** [comment]
About the suggested change in `_SeqAbstractBaseClass`, replacing (A)
```
        if isinstance(other, _SeqAbstractBaseClass):
            return self._data == other._data
        elif isinstance(other, str):
            return self._data == other.encode("ASCII")
        else:
            return self._data == other
```

with (B)

```
        if isinstance(other, _SeqAbstractBaseClass):
            return self._data == other._data
        elif isinstance(other, (bytes, bytearray)):
            return self._data == other
        else:
            return self._data == str(other).encode("ASCII")
```

(A) means that be default we treat `other` as bytes-like, and check for `str` as a special case. This works, except if `other` is not bytes-like, not a string, but should be treated as a string.
(B) means that be default we treat `other` as string-like, and check for `bytes`, `bytearray` as special cases.

I am OK with (B); it is more robust than (A).
One question though: How should we treat bytes-like objects other than `bytes`, `bytearray`?
Current behavior is
```
>>> s = Seq("ABCD")
>>> s == "ABCD"
True
>>> s == b"ABCD"
True
>>> s == bytearray(b"ABCD")
True
>>> import array
>>> t = array.array("B", b"ABCD")
>>> s == t
False
```
Is this what we want? Because the data contents are the same in each case:
```
>>> s == memoryview(t)
True
```

One more issue :with (B), any object is then equal to a `Seq` object if its string-value is the same. This may give unexpected behaviors, for example
```
>>> s = Seq("123")
>>> s == 123
True
```

**Comment by mdehoon** [comment]
@hypostulate  What is the purpose of the `SeqLike` class in your script?

**Comment by hypostulate** [comment]
@mdehoon The `SeqLike` class is a mock of the style of class I work with regularly: a class that "appears" to be a `Seq`/`str` but in reality is much more complex. The `__eq__` method is implemented quite specifically and is often used in comparison to both `Seq` and `str` objects where the equality is based on casting the `SeqLike` object to `str` before comparison. After the change to 1.79 I noticed my CI tests failing on one-sided equality against `Seq` objects due to the above scenario.

To be clear, because of the way the `__eq__` resolution works, it is solvable, there is just special handling that I needed to implement specifically for `bytes` and `bytearray` that was not necessary previously when the `Seq.__eq__` default casted the `other` to `str`. I'm not sure this warrants any change directly, so I closed the issue. Might be worth thinking about whether any other similar trickle down effects might have gone unnoticed.
